falcon/typing.py

```diff
@@ -0,0 +1,43 @@
+# Copyright 2021-2023 by Vytautas Liuolia.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Shorthand definitions for more complex types."""
+
+from typing import Any, Callable, Pattern, TypeVar, Union
+
+from falcon.request import Request
+from falcon.response import Response
+
+# Common definitions
+ExceptionClass = TypeVar('ExceptionClass', bound=BaseException)
+
+RequestClass = TypeVar('RequestClass', bound=Request)
+ResponseClass = TypeVar('ResponseClass', bound=Response)
+
+
+# Error handlers
+ErrorHandler = Callable[[RequestClass, ResponseClass, ExceptionClass, dict], Any]
+
+
+# Error serializers
+ErrorSerializer = Callable[[RequestClass, ResponseClass, ExceptionClass], Any]
+
+
+# Sinks
+# TODO(vytas): Is it possible to specify a Callable or a Protocol that defines
+#   type hints for the two first parameters, but accepts any number of keyword
+#   arguments afterwords?
+# class SinkCallable(Protocol):
+#     def __call__(sef, req: Request, resp: Response, <how to do?>): ...
```

Comment:
`**kwargs:  Any` (kwargs can be named whatever).

for any positional arg instead you can do `*args: Any` (again args can be named whatever)

Reply:
Unfortunately, that does not work for me.

Any help is very welcome on this one... until then, I'll have to leave it for better times...

This is what I tried:
```
from typing import Protocol


class FancyCallable(Protocol):
    def __call__(self, name: str, **kwargs):
        ...


def fancy_call(name: str, callback: FancyCallable):
    print(f'Hello, {name}!')
    callback(name, style='fancy', version=2.1)


def log_call(name: str, style: str = 'plain', version: float = 2.0):
    print(f'[log_call] {name=} {style=}')


def main() -> None:
    fancy_call('Federico', log_call)


if __name__ == '__main__':
    main()
```
Works perfectly just to run it, and I don't see any problems in code whatsoever.

However, `mypy` complains:
```
test_prot.py:19: error: Argument 2 to "fancy_call" has incompatible type "Callable[[str, str, float], Any]"; expected "FancyCallable"  [arg-type]
```

Reply:
I even tried forcing kw-only args on `log_call`, but it still complains...
```
test_prot.py:19: error: Argument 2 to "fancy_call" has incompatible type "Callable[[str, DefaultNamedArg(str, 'style'), DefaultNamedArg(float, 'version')], Any]"; expected "FancyCallable"  [arg-type]
```